Server-errors-and-solutions.md

```diff
@@ -0,0 +1,80 @@
+This document talks about different types of server problems that have been fixed before, and it suggests ways to stop those problems from happening again. The types of problems include issues with data, problems when lots of people are using the server, situations that don't happen very often, making sure the server works with old internet browsers, mistakes in the way the server thinks, problems when the internet is slow, and errors that don't need to be written down. For each type of problem, it gives an example and tips on how to stop it from happening.
+
+# Categories
+
+* [Scalability issue](#scalability-issue)
+* [Corner cases](#corner-cases)
+* [Supporting old browsers](#supporting-old-browsers)
+* [Logic errors](#logic-errors)
+* [Data related issue](#data-related-issues)
+* [Race conditions / issues on slow networks](#race-conditions)
+* [Related to migration](#related-to-migrations)
+* [Errors that didn't need to be logged](#errors-that-didn't-need-to-be-logged)
+
+## Scalability Issue
+
+Issues that are related to features not scaling well on production. Example: (1) Exploration state data is too large(#12547) for some explorations and no further changes can be made to these explorations. (2) Suggestion models being fetched for review exceeds the query limit, so only an incomplete result is shown.
+
+### Preventive action
+
+* __Design stage__: During design doc reviews, one should ask questions about performance and scalability when dealing with large amounts of data – “how big do you expect X to get?”, “are there any limits to the size?”, “what happens when it reaches this limit and is it possible to handle it?”
+* __Manual testing__: Having a functionality to generate dummy data for testing locally might help.
+
+## Corner Cases
+
+User-facing issues where certain corner case scenarios are not considered (empty list, undefined object etc.) Example: Deleting a hint that is active in the editor view caused console errors. Although these are not often serious, these issues can be tricky to debug due to lack of reproduction steps. The majority of server errors fall under this category. Other e.g. error #7907, #6298, #6199, #12528, #8702, #8464, #7970, #7546, #7444, #7727 (usage of plurals in certain languages), #7342, #7331, #7047, #6757, #6523 (didn't account for changes to interaction type), #5457
```

Comment:
Make sure your links work (they currently don't). Use Markdown format for links.

Also, you don't need so many examples, at most 5 will do.